text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,559 @@
+an extracted context is not intended to be set as current context. An 
+alternate extract method can be added to the API to handle this.
+
+```php
+// Most of the time, the extract function operates on the current context.
+Extract(headers)
+
+// When a context needs to be extracted without changing the current 
+// context, fall back to the explicit API.
+otherContext = ExtractWithContext(Context::GetCurrent(), headers)
+```
+
+# Internal details
+
+![drawing](img/context_propagation_details.png)
+
+## Context details
+OpenTelemetry currently intends to implement three context types of context 
+propagation.
+
+**Span Context -** The serializable portion of a span, which is injected and 
+extracted. The readable attributes are defined to match those found in the 
+[W3C Trace Context specification](https://www.w3.org/TR/trace-context/). 
+
+**Correlation Context -** Correlation Context contains a map of labels and 
+values, to be shared between metrics and traces. This allows observability data 
+to be indexed and dimensionalized in a variety of ways. Note that correlations 
+can quickly add overhead when propagated in-band. But because this data is 
+write-only, it may be possible to optimize how it is transmitted.
+
+**Baggage -** Transaction-level application data, meant to be shared with 
+downstream components. This data is readable, and must be propagated in-band. 
+Because of this, Baggage should be used sparingly, to avoid ballooning the size 
+of all downstream requests.
+
+Note that OpenTelemetry APIs calls should *always* be given access to the entire 
+context object, and never just a subset of the context, such as the value in a 
+single key. This allows the SDK to make improvements and leverage additional 
+data that may be available, without changes to all of the call sites.
+
+
+## Context Management and in-process propagation
+
+In order for Context to function, it must always remain bound to the execution 
+of code it represents. By default, this means that the programmer must pass a 
+Context down the call stack as a function parameter. However, many languages 
+provide automated context management facilities, such as thread locals. 
+OpenTelemetry should leverage these facilities when available, in order to 
+provide automatic context management.
+
+## Pre-existing Context implementations
+
+In some languages, a single, widely used Context implementation exists. In other 
+languages, there many be too many  implementations, or none at all. For example, 
+Go has a the `context.Context` object, and widespread conventions for how to 
+pass it down the call stack.
+
+In the cases where an extremely clear, pre-existing option is not available, 
+OpenTelemetry should provide its own Context implementation.
+
+## Default Propagators
+
+When available, OpenTelemetry defaults to propagating via HTTP header 
+definitions which have been standardized by the W3C.
+
+
+# FAQ
+
+## Why separate Baggage from Correlations?
```

Comment:
Personally, I find the arguments for splitting to be very hand-wavy at the moment, especially the reference to some magical "optimizations, such as propagating the correlation data out-of-band" (which doesn't make much sense for metrics). The delivery guarantee seems to be the only real reason for the split. However, this could be achieved with a single concept of baggage where each item has a priority.

@bogdandrutu didn't you mention last week at W3C workshop that you were also having second thoughts about the split?

Reply:
@yurishkuro there has been very consistent feedback in OpenTracing that Observability implementations should not be required to carry baggage.

Another example of the split: sampling and disabled observability. When tracing/metrics is a noop, baggage must still be propagated. So the propagation rules for correlations and baggage are different, even if the 

There is an argument that some baggage values should also be available as correlations. One option is to just add these values to both - honestly I don't see this as a huge issue. Another option is that the OTel SDK could also implement the baggage API, in addition to Tracing and metrics. So the APIs are separate, which simplifies the propagation rules for users. But under the hood we can use all the data.

Reply:
> there has been very consistent feedback in OpenTracing that Observability implementations should not be required to carry baggage.

Which is why context and propagation in OpenTelemetry are separate and have nothing to do with observability. Observability aspects __use__ the propagation, they do not __provide it__.

Reply:
I agree that this section could make a better argument.

But, at this point, there is long line of OpenTelemetry member organizations which have requested that baggage be removed from tracing. @yurishkuro, I don't feel like you last comment addressed their concerns. 

Remember, Baggage is not Context Propagation. Span context is not baggage, for example. Spans, Correlations, and Baggage are three separate uses of Context Propagation. 

I'd rather have a clean separation of concerns in the API, and then find a way for the SDK to optimize under the hood, should we desire to do so.

 I find the arguments for keeping baggage within the tracer to be flimsy:
**A)** sending the value twice creates overhead.
**B)** users may add a baggage item, and the metrics system may want to grab the value even though the user did not specify they wanted this to be used as a correlation.
**C)** Setting a value twice is annoying.

There are answers to all of these:
**A)** The SDK can optimize this, should we want it to.
**B)** This is just sloppy. And what if the user did not want baggage to be recorded?
**C)** It is easy to offer a convenience method to do both.

If we want to clean up our context propagation, but leave baggage and correlations together... okay... but I think both you @bogdandrutu are forgetting how much pain and confusion the current situation is causing people today.

Reply:
This comment by @yurishkuro is :100:!

> Observability aspects use the propagation, they do not provide it.

This deserves a bit of explicit attention at the top of the document. The point is that these are only decoupled in a semantic sense. The SDK will surely depend on some concrete behavior of the propagation system if we are to benefit from in-band correlation context propagation. 

The user should not be able to access the propagation aspect, only the SDK should.

Reply:
I think correlations differs from baggage in important ways, particularly around the write-only nature of correlations.  One of the key distinctions about correlations is that an SDK should be permitted to drop them of there are too many--this will not affect the application because correlations are write-only by definition. Dropping correlations can only degrade the observability system, by definition.

Baggage values cannot be dropped in this way, as they are application state by definition.

IMO the goal of decoupling propagation from other observability aspects is to define semantics to be shared by the observability functions that may use correlations. There will be a default SDK for metrics and tracing that uses in-band correlation propagation, these default SDKs will depend on a default propagation API which serves in-band propagation.

We've discussed a streaming SDK for tracing and/or metrics that simply records events as they happen without keeping state in the process. A streaming propagation SDK could be coupled with streaming metrics and tracing SDKs that simply records new correlation-context events. The downstream system should be able to reconstruct the correlation state of every downstream span and metric event, according to the semantics of the recorded propagation events, with no requirement for in-band propagation.

Baggage, on the other hand, must be propagated in-band for the user is allowed to read it.